src/algorithms/base_algorithm.cc

```diff
@@ -119,6 +119,7 @@ void BaseAlgorithm::initialize() {
     mixing->set_num_components(data.rows());
   }
   // Initialize hierarchies
```

Comment:
We are now calling `set_dataset` from `base_algorithm.cc` but we only implemented it in `MFAHierarchy`, since we didn't know if and how you want to handle this new `dataset_ptr` member.

Reply:
I think it's a good idea to have the dataset_ptr in all the hierarchies, you can move the pointer and setter to the BaseHierarchy class :)